src/util/vecdeque.h

```diff
@@ -0,0 +1,271 @@
+                    std::memcpy(new_buffer, m_buffer + m_offset, first_part * sizeof(T));
+                }
+                if (first_part != m_size) {
+                    std::memcpy(new_buffer + first_part, m_buffer, (m_size - first_part) * sizeof(T));
+                }
+            } else {
+                // Otherwise move-construct in place in the new buffer, and destroy old buffer objects.
+                size_t old_pos = m_offset;
+                for (size_t new_pos = 0; new_pos < m_size; ++new_pos) {
+                    std::construct_at(new_buffer + new_pos, std::move(*(m_buffer + old_pos)));
+                    std::destroy_at(m_buffer + old_pos);
+                    ++old_pos;
+                    if (old_pos == m_capacity) old_pos = 0;
+                }
+            }
+        }
+        // Deallocate old buffer and update housekeeping.
+        std::allocator<T>().deallocate(m_buffer, m_capacity);
+        m_buffer = new_buffer;
+        m_offset = 0;
+        m_capacity = capacity;
+        Assume((m_offset == 0 && m_capacity == 0) || m_offset < m_capacity);
+    }
+
+    /** What index in the buffer does logical entry number pos have? */
+    size_t BufferIndex(size_t pos) const noexcept
+    {
+        if (pos >= m_capacity - m_offset) {
+            return pos - (m_capacity - m_offset);
+        } else {
+            return pos + m_offset;
+        }
+    }
+
+    /** Specialization of resize() that can only shrink. Separate so that clear() can call it
+     *  without requiring a default T constructor. */
+    void ResizeDown(size_t size) noexcept
+    {
+        Assume(size <= m_size);
+        if constexpr (std::is_trivially_destructible_v<T>) {
+            // If T is trivially destructible, we do not need to do anything but update the
+            // housekeeping record. Default constructor or zero-filling will be used when
+            // the space is reused.
+            m_size = size;
+        } else {
+            // If not, we need to invoke the destructor for every element separately.
+            while (m_size > size) {
+                std::destroy_at<T>(m_buffer + BufferIndex(m_size - 1));
+                --m_size;
+            }
+        }
+    }
+
+public:
+    VecDeque() noexcept = default;
+
+    void resize(size_t size)
+    {
+        if (size < m_size) {
+            // Delegate to ResizeDown when shrinking.
+            ResizeDown(size);
+        } else if (size > m_size) {
+            // When growing, first see if we need to allocate more space.
+            if (size > m_capacity) Reallocate(size);
+            while (m_size < size) {
+                std::construct_at(m_buffer + BufferIndex(m_size));
+                ++m_size;
+            }
+        }
+    }
+
+    void clear() noexcept { ResizeDown(0); }
+
+    ~VecDeque()
+    {
+        clear();
+        Reallocate(0);
+    }
+
+    VecDeque& operator=(const VecDeque& other)
+    {
+        if (&other == this) return *this;
```

Comment:
Does it make sense to label branches with the `[[likely]]` and `[[unlikely]]` attributes here?